test/infrastructure/docker/internal/webhooks/devmachine_webhook.go

```diff
@@ -0,0 +1,81 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+)
+
+// DevMachine implements a validating and defaulting webhook for DevMachine.
+type DevMachine struct{}
+
+func (webhook *DevMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&infrav1.DevMachine{}).
+		WithDefaulter(webhook).
+		WithValidator(webhook).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-infrastructure-cluster-x-k8s-io-v1beta1-devmachine,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=devmachines,versions=v1beta1,name=default.devmachine.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+
+var _ webhook.CustomDefaulter = &DevMachine{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+func (webhook *DevMachine) Default(_ context.Context, obj runtime.Object) error {
+	machine, ok := obj.(*infrav1.DevMachine)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a DevMachine but got a %T", obj))
+	}
+	defaultDevMachineSpec(&machine.Spec)
+	return nil
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1beta1-devmachine,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=devmachines,versions=v1beta1,name=validation.devmachine.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+
+var _ webhook.CustomValidator = &DevCluster{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *DevMachine) ValidateCreate(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *DevMachine) ValidateUpdate(_ context.Context, _, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *DevMachine) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func defaultDevMachineSpec(s *infrav1.DevMachineSpec) {
+	if s.Backend.InMemory == nil &&
```

Comment:
Do we need something similar for the other objects?

Reply:
I hope I get the feedback right, PTAL

Reply:
I think partially :)

So I was wondering if we are defaulting here to use docker if no backend is specific, what about other objects.
You implemented it now also for DevMachineTemplate.

Do we need the same for DevCluster and DevClusterTemplate?

Reply:
It is already done also for  DevCluster and DevClusterTemplate